Readme.md
# mongo-query-exporter

Prometheus экспортёр, который выполняет запросы в mongodb и возвращает метрики.


## Сборка 

Для зависимостей используется [go-dep](https://github.com/golang/dep)

        dep ensure
        go build -o mongo-query-exporter  main.go collector.go

## Запуск 

                ./mongo-query-exporter --help 
                usage: mongo-query-exporter --port=PORT [<flags>]

                Flags:
                --help           Show context-sensitive help (also try --help-long and --help-man).
                --port=PORT      Prometheus exporter listen port
                --config=CONFIG  Prometheus exporter listen port
                --version        Show application version.


`--port` обязательный аргумент

`--config` по умолчанию используется config.yaml, можно указать путь к другому файлу.


## Конфигурация

### Пример
```Yaml

collector_name: mongo-dev
database_uri: mongodb://mongo.dev/
metrics:
  - metric_name: esia_users # Название метрики
    type: gauge # Не импользуется
    function: aggregate # Не импользуется
    database: user # База данных
    collection: user # Коллекция
    interval: 5s # Время жизни кэша. После выполнения запроса он попадает в кэш на время interval. 
    value: count # Значение. `{ "_id" : null, "count" : 20 }`
    query: |
      [ {$match: {$and:[{"some_user_id": {$ne: null}}, { $or: [{"some_date": {$lte: "2019-02-28 23:59:59"} },{"some_date": null}]}]}}, { $group: { _id: "$id", count: {$sum:1} } }]
```

interval - в формате golang time.Duration. Например можно использовать:  `1s,1m,1h,1d`

### Дата/Время

Поддерживается указание даты в форматах:

```js 
(new Date("2016-01-01 23:59:59"))
ISODate("2019-01-01 23:59:59")
```
Указанные даты парсятся, конвертируюся в EXT JSON в формате `{$date:{$numberLong:"134234234234"}}`, далее преобразуются в bson запрос.

CHANGELOG.md
# v0.0.2 (2020-01-20)











collector.go
package main

import (
	"context"
	"io/ioutil"
	"log"
	"reflect"
	"time"

	"github.com/patrickmn/go-cache"
	"github.com/prometheus/client_golang/prometheus"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"gopkg.in/yaml.v2"
)

//Define a struct for you collector that contains pointers
//to prometheus descriptors for each metric you wish to expose.
//Note you can also include fields of other types if they provide utility
//but we just won't be exposing them as metrics.

type Querie struct {
	// Запрос
	name       string
	metric     *prometheus.Desc
	db         string
	collection string
	q          []bson.D
	value      string
	function   string
	interval   time.Duration
}

type Collector struct {
	// Коллектор
	queries []*Querie
	client  *mongo.Client
	ctx     context.Context
	cache   *cache.Cache
}

func NewCollector() *Collector {
	// Инициирем метрики
	var configFile string
	if *config == "" {
		configFile = "config.yaml"
	} else {
		configFile = *config
	}

	buf, err := ioutil.ReadFile(configFile)
	if err != nil {
		log.Fatal(err)
	}

	conf := Config{}
	err = yaml.Unmarshal(buf, &conf)
	if err != nil {
		log.Fatal("Config parsing error")
	}
	ctx := context.Background()
	client := Connect(ctx, conf.DatabaseURI)

	// Инициируем кэш
	c := cache.New(cache.NoExpiration, cache.NoExpiration)

	MongoCollector := &Collector{}
	MongoCollector.client = client
	MongoCollector.ctx = ctx
	MongoCollector.cache = c

	q := &Querie{}
	for _, m := range conf.Metrics {
		q = &Querie{
			name:       m.MetricName,
			metric:     prometheus.NewDesc(m.MetricName, m.Query, []string{"field"}, nil),
			db:         m.Database,
			collection: m.Collection,
			q:          Normalize(m.Query),
			value:      m.Value,
			function:   m.Function,
			interval:   m.Interval,
		}
		MongoCollector.queries = append(MongoCollector.queries, q)

	}

	return MongoCollector
}

//Each and every collector must implement the Describe function.
//It essentially writes all descriptors to the prometheus desc channel.
func (collector *Collector) Describe(ch chan<- *prometheus.Desc) {

	for _, m := range collector.queries {
		ch <- m.metric
	}

}

//Collect implements required collect function for all promehteus collectors
func (collector *Collector) Collect(ch chan<- prometheus.Metric) {
	// Берём метрики из канала, получаем для них данные

	c := collector.cache
	// log.Println("ItemCount: ", c.ItemCount())
	for _, m := range collector.queries {
		var result interface{}
		// var s reflect.Value
		// log.Println(m.name)
		// Проверяем наличие метрики в кэше
		CachedValue, found := c.Get(m.name)
		if found {
			// fmt.Println("Get from cache: ", CachedValue)
			result = CachedValue
			// s = reflect.ValueOf(result)

		} else {
			// Добавляем метрику в кэш на время interval
			// fmt.Println("Cache not found, making querie")
			result = GetAggregate(collector.ctx, collector.client, m.db, m.collection, m.q, m.value)
			c.Set(m.name, result, m.interval)
			// s = reflect.ValueOf(result)
		}

		// log.Println(s.Len())
		mlist := result.([]map[string]interface{})
		if len(mlist) == 1 {
			v := reflect.ValueOf(mlist[0][m.value]).Int()
			ch <- prometheus.MustNewConstMetric(m.metric, prometheus.GaugeValue, float64(v), "empty")

		} else {
			for _, j := range mlist {

				v := reflect.ValueOf(j[m.value]).Int()
				label := reflect.ValueOf(j["_id"]).String()
				ch <- prometheus.MustNewConstMetric(m.metric, prometheus.GaugeValue, float64(v), label)
			}
		}

	}

}


main.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"regexp"
	"strings"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/robertkrimen/otto"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"gopkg.in/alecthomas/kingpin.v2"
)

type Config struct {
	// Конфиг файл
	CollectorName string `yaml:"collector_name"`
	DatabaseURI   string `yaml:"database_uri"`
	Metrics       []struct {
		MetricName string        `yaml:"metric_name"`
		Type       string        `yaml:"type"`
		Function   string        `yaml:"function"`
		Database   string        `yaml:"database"`
		Collection string        `yaml:"collection"`
		Interval   time.Duration `yaml:"interval"`
		Value      string        `yaml:"value"`
		Query      string        `yaml:"query"`
	} `yaml:"metrics"`
}

var (
	port   = kingpin.Flag("port", "Prometheus exporter listen port").Required().Int()
	config = kingpin.Flag("config", "Prometheus exporter listen port").String()
)

func main() {
	kingpin.Version("0.0.1")
	kingpin.Parse()
	c := NewCollector()
	prometheus.MustRegister(c)

	http.Handle("/metrics", promhttp.Handler())
	log.Println("Beginning to serve on port :", *port)
	log.Fatal(http.ListenAndServe(strings.TrimSpace(fmt.Sprintln(":", *port)), nil))
}

func Normalize(query string) []bson.D {
	// Преобразует js -> json ->bson
	DateRegexp := regexp.MustCompile(`\(?(new\sDate|ISODate)\(.(\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2}).\)\)?`)
	match := DateRegexp.FindAllStringSubmatch(query, -1)

	// Преобразуем (new Date("2016-01-01 23:59:59")) -> {$date:{$numberLong:"134234234234"}}
	if match != nil {
		log.Println("Trying to convert time: ", match)
		for _, i := range match {
			t, err := time.Parse("2006-01-02 15:04:04", i[2])

			if err != nil {
				log.Fatalln(err, i[2])
			}
			query = strings.Replace(query, i[0], fmt.Sprintf(`{$date:{$numberLong:"%v"}}`, int64(t.Unix()*1000)), 1)

		}
	}

	// Преобразуем js в json
	vm := otto.New()
	jsvalue, err := vm.Run(query)
	output, err := jsvalue.Export()
	// log.Println(fmt.Sprintln(output))
	output2, err := json.Marshal(output)
	// Преобразуем extjson в bson
	bdoc := make([]bson.D, 0)
	err = bson.UnmarshalExtJSON([]byte(output2), true, &bdoc)
	if err != nil {
		log.Fatal("UnmarshalExtyaml: ", err)
	}
	// for _, i := range bdoc {
	// 	log.Println(i.Map())

	// }

	log.Println("BSON query: ", bdoc)
	return bdoc

}

func Connect(ctx context.Context, uri string) *mongo.Client {
	// Подключаемся к базе
	log.Println("Trying connect to database, uri: ", uri)
	client, err := mongo.NewClient(options.Client().ApplyURI(uri))
	if err != nil {
		log.Fatal(err)
	}
	err = client.Connect(ctx)
	if err != nil {
		log.Fatal(err)
	}
	// Check the connection
	err = client.Ping(ctx, nil)
	if err != nil {
		log.Fatal(err)
	}
	log.Println("Connected to MongoDB!")
	return client
}

func GetAggregate(ctx context.Context, client *mongo.Client, db string, coll string, query []bson.D, value string) interface{} {
	// Выполняем запрос с функцией aggregate
	collection := client.Database(db).Collection(coll)
	// log.Println(collection.Name())
	res, err := collection.Aggregate(ctx, query)
	if err != nil {
		log.Fatal(err)
	}
	var result []map[string]interface{}
	err = res.All(ctx, &result)
	return result

}
